### ISSUE TITLE
[Community Discussion] Progressive Tasks Datasets with Verification for Agentic RL

### ISSUE THREAD

**Issue by August-murr** [huggingface/trl#3417]
I was hoping this would kick off a project where we start gathering data for Agentic RL, but there are some hurdles we need to talk about. I’d love for everyone to weigh in and share their thoughts.

Agentic RL needs a different kind of dataset compared to other methods, specifically **Verifiable Tasks**. Most of the available datasets out there are like LeetCode-style coding problems, such as [LeetCodeDataset](https://huggingface.co/datasets/newfacade/LeetCodeDataset). 

While those are good for demos, they don’t have any real-world applications.

**Let’s discuss some of these challenges.**

### Integration with environments
The development of environments has been speeding up since they can easily connect with APIs and aren't necessarily used for training. The challenge with them in agentic reinforcement learning is that, unlike datasets like GSM8K or HotpotQA, which have clear final answers to compare with, agentic RL needs tasks and their verification to be specifically tailored for each environment. For example, if the agent is tasked with writing a function to solve a problem, the context and verification would look pretty different if you’re using a sandboxed code executor versus a repository or MCP.

### Cost of manually written verifications
It's arguably the slowest and most expensive part of Agentic RL because you have to create a custom verifier for every task, which is a problem because of the human in the loop. One idea worth exploring to make verification easier is self-evaluation. Maybe the agent can get better at judging its own performance as it improves at the tasks, or it could even be trained specifically for that.

### General vs. Specialized Tasks datasets
In the same way we had SFT and preference datasets, both general and specialized for specific tasks, we can choose to curate:

1. A General Tasks Dataset aimed at AGI.
2. A Specialized Tasks Dataset focused on practical uses, like assistance or software engineering.

I believe we should aim at automating AI research. We're not gonna build AGI, the Autonomous AI Researcher will.

### AI research and engineering tasks dataset
think of any tasks from entry-level ML tasks all the way to tasks done by an entire research organization, the majority of which require lots of GPUs, which is why I've already written a sandboxed [code executer with GPUs](https://github.com/August-murr/AutonomousAIResearcher/pull/3) for such tasks. though the use case is very limited due to the limit of sandboxes and tasks datasets.

### Progressive Dataset for sparse reward signals
A progressive dataset involves labeling each task with a difficulty level, which can ideally be set by the agent being trained. This label is measured by the success rate over n attempts. The idea is to train the agent only on tasks that give some reward signal, avoiding tasks that are too easy (like those with a success rate of 90% or higher) and those that are too difficult to get any reward signals from (0%).
It assumes the knowledge and expertise gained from some tasks transfer to other tasks and helps provide reward signals for more difficult tasks.

Hopefully, after discussing and deciding on which environments to use, we can begin collecting and writing different types of Tasks datasets and train agents.

**Comment by August-murr** [comment]
Assuming that creating the tasks dataset really depends on the environment we’re working in, the most flexible setup we could come up with right now could be something like a fork of [OpenAI Codex CLI](https://github.com/openai/codex) that connects to with the model we’re training. This setup would let the agent read, write, and run things on sandboxed instances like E2B, and if we need some GPU, we could use Modal sandboxes. We’d also include the necessary dependencies and the codebase that the agent will be working with. Then, we’d create a Tasks Dataset with test cases that run in the same sandbox used by the Agent, and we’d collect the rollout for the training step.

example:
**Task**:
"You are Qwen and you are hosted at 'Qwen/Qwen2.5-Coder-7B-Instruct'. fine-tune yourself on this medical dataset: 'FreedomIntelligence/medical-o1-reasoning-SFT' and save the fine-tuned model at 'path/to/save'. you have access to 4 H100 GPUs, which costs 0.27$/Hour and you have a budget of 1$."

**Test Case:**
```python
from transformers import AutoModelForCausalLM
from datasets import load_dataset
from evaluate import evaluate
model = AutoModelForCausalLM.from_pretrained('path/to/save')
eval_dataset = load_dataset("FreedomIntelligence/medical-o1-reasoning-SFT",split="test")
evaluator = evaluate("text-classification")
result = evaluator.compute(model,eval_dataset)
print(result)
```

the reward would be calculated with a combination of the eval metric and the cost of the sandboxed GPU instances.
